src/fireedge/src/modules/utils/helpers.js

```diff
@@ -0,0 +1,758 @@
+  const cache = new Map()
+
+  /**
+   * @param {object} root0 - Params
+   * @param {object} root0.obj - Object to search
+   * @param {string} root0.keyToFind - Key to find
+   * @param {Array} root0.path - Path to key
+   * @param {boolean} root0.findAll - Return first or all
+   * @returns {object} - Found key(s) + path(s)
+   */
+  const search = ({ obj, keyToFind, path = [], findAll = false }) => {
+    const cacheKey = JSON.stringify({ obj, keyToFind, path, findAll })
+    if (cache.has(cacheKey)) {
+      return cache.get(cacheKey)
+    }
+
+    if (typeof obj !== 'object' || obj === null) {
+      return { found: false, paths: [] }
+    }
+
+    if (keyToFind in obj) {
+      const newPath = [...path, keyToFind]
+      const searchResult = { found: true, paths: [newPath] }
+      if (!findAll) {
+        cache.set(cacheKey, searchResult)
+
+        return searchResult
+      }
+
+      return searchResult
+    }
+
+    let accumulatedResults = []
+
+    const entries = Array.isArray(obj) ? obj.entries() : Object.entries(obj)
+    for (const [key, value] of entries) {
+      if (typeof value === 'object') {
+        const nextPath = Array.isArray(obj)
+          ? [...path, `[${key}]`]
+          : [...path, key]
+        const subSearchResult = search({
+          obj: value,
+          keyToFind,
+          path: nextPath,
+          findAll,
+        })
+        if (subSearchResult.found) {
+          accumulatedResults = [...accumulatedResults, ...subSearchResult.paths]
+          if (!findAll) {
+            break
+          }
+        }
+      }
+    }
+
+    const found = accumulatedResults.length > 0
+    const finalResult = { found, paths: accumulatedResults }
+    cache.set(cacheKey, finalResult)
+
+    return finalResult
+  }
+
+  return search
+})()
+
+/**
+ * Check if data is an array and if it's only one element, return an array with one element.
+ *
+ * @param {object} data - Object or list
+ * @returns {Array} An array with one element or the same array
+ */
+export const responseDataToArray = (data) =>
+  data ? (Array.isArray(data) ? data : [data]) : undefined
+
+/**
+ *
+ * @param {object} OpennebulaObject - OpennebulaObject
+ * @returns {string} - If OpennebulaObject is locked/unlocked
+ */
+export const getLocked = (OpennebulaObject) => !!+OpennebulaObject.LOCK?.LOCKED
+
+/**
+ * @param {string} errorId - Unique error identifier
+ * @param {object} root0 - Options
+ * @param {string} root0.fallback - Optional fallback error message
+ * @returns {string} - Formatted error message
+ */
+export const formatError = (errorId = '', { fallback = '' } = {}) =>
+  sentenceCase(ERROR_LOOKUP_TABLE?.[String(errorId) || ''] ?? fallback)
+
+/**
+ * @param {object} obj - Dirty object
+ * @returns {object} - Object with falsy values removed
+ */
+export const deepClean = (obj) => {
+  if (_.isArray(obj)) {
+    return obj.map(deepClean)
+  } else if (_.isObject(obj)) {
+    return _.pickBy(
+      _.mapValues(obj, deepClean),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This route is missing authorization checks. Without proper authorization, sensitive routes may be exposed to unauthorized users, leading to potential data breaches or unauthorized actions. Ensure that authorization middleware or handlers are applied. For more information on missing authorization vulnerabilities, refer to [CWE-862](https://cwe.mitre.org/data/definitions/862.html).